distributed/scheduler.py

```diff
@@ -8253,19 +8268,61 @@ async def get_worker_monitor_info(self, recent=False, starts=None):
     # Cleanup #
     ###########
 
-    async def check_worker_ttl(self):
+    @log_errors
+    async def check_worker_ttl(self) -> None:
         now = time()
         stimulus_id = f"check-worker-ttl-{now}"
+        to_restart = {}
+        to_close = set()
+        assert self.worker_ttl
+        ttl = max(self.worker_ttl, 10 * heartbeat_interval(len(self.workers)))
+
         for ws in self.workers.values():
-            if (ws.last_seen < now - self.worker_ttl) and (
-                ws.last_seen < now - 10 * heartbeat_interval(len(self.workers))
-            ):
+            last_seen = now - ws.last_seen
+            if last_seen > ttl:
+                if ws.nanny:
+                    to_restart[ws.address] = ws
+                    msg = "restarting"
+                else:
+                    to_close.add(ws)
+                    msg = "closing"
+                # In case the worker resuscitates while we're restarting it, prevent
+                # it from receiving tasks from AMM
+                self.handle_worker_status_change(
+                    Status.failed, ws, stimulus_id=stimulus_id
+                )
                 logger.warning(
-                    "Worker failed to heartbeat within %s seconds. Closing: %s",
-                    self.worker_ttl,
-                    ws,
+                    f"Worker failed to heartbeat for {last_seen} seconds; {msg}: {ws}"
                 )
-                await self.remove_worker(address=ws.address, stimulus_id=stimulus_id)
+
+        if to_restart:
+            result = await self.broadcast(
+                msg={"op": "restart", "reason": stimulus_id},
```

Comment:
out of scope: I'm not a fan of the `reason` parameter; we should replace it everywhere with `stimulus_id`.